### PR TITLE
Fancy IPA Library + New Downloader + New Cache System

### PR DESCRIPTION
**Pull request by amirsaam**
Co-Authored-By: Isaac Marovitz <42140194+IsaacMarovitz@users.noreply.github.com>

This PR introduces several changes:

- New detailed page for each app in IPA Libraey based on iTunesLookup file that its function is already being used and handles fetching app icons in IPA Library (No extra function is required) by @IsaacMarovitz
- New `StackNavigationView` view extension has been introduced in `ViewExtenstions.swift` that is similar to `NavigationStack` on macOS 13 but supports macOS 12, so DetailedView is accessible in macOS 12 with same code. This extension can be used in any other views as you may want like creating App Folders in App Library or Source Folders in IPA Library
- New `DownloadApp` class has been introduced in `Downloader.swift` explained [here](https://github.com/PlayCover/PlayCover/pull/588#issuecomment-1361166253)
- Added working Download or Cancel Download buttons in the detailed page of each app
- Download buttons will show 4 different text:
1. `Get`: If the user hasn't installed that app/game yet on their PlayCover
2. `Reinstall`: If the user already has the same version of the app installed in their PlayCover
3. `Update`: If user has older version compared to what they are viewing installed in their PlayCover
4. `Downgrade`: If user has newer version compared to what they are viewing installed in their PlayCover

Note: This version compare is using same logic and function of what @TheMoonThatRises has introduced into PlayCover in previously merged PRs.

- Download buttons on press will change to Download progress bar with a stop option and after download to Install progress bar and will be shown only when user is in that app detail page
- Made showing ToastView optional that can be controlled via `ToastVM.shared.isShown` boolean, it will no longer be shown when downloading or installing an app while user is in that app's DetailedView

Note: If you navigate to other apps DetailedView you'll still see the 4 type of download texts but they are disabled, no matter what in any DetailedView while app is being installed, that button is disabled.

- Apps that have no iTunesLookup will get default text that notifies user the app is not on app store
- New `Cacher` class has been introduced in `Cacher.swift` that uses `DataCache` package, removed `ImageCache` class and rewrote the logic.
- Online Images now are being cached via `CachedAsyncImage` package that uses `URLCache` as memory 
- All texts have locale key, so translation can start right away

This library in future PlayCover development can be used to shown extra details that can be provided by sources like a stat badge like file size that shows if the app should have PlayTools or not or it just works in SIP disabled environment or not

Overall the general point is, current IPA Library interface has no room for improvement and showing details that user need to be notified about apps in sources.

Demo: [this comment](https://github.com/PlayCover/PlayCover/pull/588#issuecomment-1363776097)

### CODE REVIEW

**Comment by JoseMoreville** [comment]
This will be through a review by all the other devs to see if it is viable to be implemented.

**Comment by amirsaam** [comment]
Demo:

https://user-images.githubusercontent.com/1331289/208268750-aa543f34-a820-4840-95f3-ca8b09d70904.mp4

**Comment by amirsaam** [comment]
Updated PR's description to show what it actually does.

**Comment by amirsaam** [comment]
A new DownloadApp class has been introduced into PlayCover.
Able to cancel downloads, retries download as how many time you config it to do (default is 3)

Note: The DownloadManager package can handle multiple downloads at the same time and cancel each one of them if user wants but I avoided to implement that into PlayCover for multiple reasons like Installer.

**Comment by amirsaam** [comment]
DownloadApp class usage example:
```swift
let downloader = DownloadApp(url: DownloadURL,
                             app: StoreAppDataObject,
                             warning: VersionCompareWarning)
if downloadVM.downloading {
    downloader.cancel()
} else {
    downloader.start()
}
```

**Comment by amirsaam** [comment]
New Downloader demo:

https://user-images.githubusercontent.com/1331289/208897959-33b7c79e-85d4-4d27-a5fd-e610eebec9ce.mp4

**Comment by amirsaam** [comment]
Last demo, no more changes on my side

https://user-images.githubusercontent.com/1331289/208950609-dbb98ac5-c9d8-4e9e-a761-16a018ff525a.mp4

**Comment by amirsaam** [comment]
Demo: 

https://user-images.githubusercontent.com/1331289/209309328-4496bd12-9bee-435c-b61b-b8bd9302033c.mp4

**Comment by JoseMoreville** [comment]
Please separate into multiple PR if possible

**Comment by amirsaam** [comment]
> Please separate into multiple PR if possible

Unfortunately some of works that have been done are because of the first intention, separating requires a rework, if the final decision for the ipa library part be drop, i would gladly rework downloader and cache part for current system.
I sincerely like you to understand 👍

File: PlayCover/AppInstaller/Downloader.swift [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    func cancel() {
+        downloader.cancelAllDownloads()
+        dlVM.downloading = false
+        dlVM.progress = 0
+        dlVM.storeAppData = nil
+    }
+
+    private func proceedDownload() {
+        let path = NSURL.fileURL(withPathComponents: [NSTemporaryDirectory(), app.bundleID])
```
Comment (IsaacMarovitz) [inline PlayCover/AppInstaller/Downloader.swift:72]: This should use the FileManager API not NSURL

File: PlayCover.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -951,12 +975,36 @@
...
 		};
-		6EC73F38293FCBD60047A07C /* XCRemoteSwiftPackageReference "inject" */ = {
+		68C7649A295478010017F054 /* XCRemoteSwiftPackageReference "inject" */ = {
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/paradiseduo/inject";
 			requirement = {
-				branch = master;
-				kind = branch;
+				kind = upToNextMajorVersion;
+				minimumVersion = 1.0.0;
```
Comment (IsaacMarovitz) [inline PlayCover.xcodeproj/project.pbxproj:983]: Changes to inject are out of scope for this PR

File: PlayCover/AppInstaller/Downloader.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    let app: StoreAppData
+    let warning: String?
+
+    init(url: URL, app: StoreAppData, warning: String?) {
+        self.url = url
+        self.app = app
+        self.warning = warning
+    }
+
+    let dlVM = DownloadVM.shared
```
Comment (IsaacMarovitz) [inline PlayCover/AppInstaller/Downloader.swift]: Avoid declaring these types of constants it makes it harder for other people to understand or refactor your code. At the very least these variable names should be changed to `downloadVM` and `installVM`

File: PlayCover/Utils/ViewExtenstions.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            content
+                .textCase(.uppercase)
+                .font(.subheadline)
+                .bold()
+                .foregroundColor(Color(nsColor: .tertiaryLabelColor))
+        }
+    }
+}
+
+struct VerticalSpacer: View {
```
Comment (IsaacMarovitz) [inline PlayCover/Utils/ViewExtenstions.swift]: This struct should be in the same swift file as the one where it is used. It is a nice shorthand used in one view, not a global utility.

File: PlayCover/Views/App Views/DetailStoreAppView.swift [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        .onDisappear {
+            ToastVM.shared.isShown = true
+        }
+        .onChange(of: downloadVM.storeAppData) { _ in
+            if downloadVM.storeAppData == app {
+                ToastVM.shared.isShown = false
+            }
+        }
+    }
+    func versionCompare() async {
```
Comment (IsaacMarovitz) [inline PlayCover/Views/App Views/DetailStoreAppView.swift:229]: Spacing

File: PlayCover/Views/App Views/DetailStoreAppView.swift [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                warningMessage = "ipaLibrary.version.same"
+            case .orderedDescending:
+                dlButtonText = "ipaLibrary.detailed.dlnewer"
+                warningMessage = "ipaLibrary.version.newer"
+            default:
+                warningMessage = "ipaLibrary.download"
+            }
+        }
+    }
+    func getData() async {
```
Comment (IsaacMarovitz) [inline PlayCover/Views/App Views/DetailStoreAppView.swift:247]: Spacing

File: PlayCover/Views/App Views/StoreAppView.swift [hunk 7]
```diff
@@ -19,24 +21,58 @@ struct StoreAppView: View {
...
-                                selected: $selected,
-                                app: app,
-                                isList: isList,
-                                warningSymbol: $warningSymbol,
-                                warningMessage: $warningMessage)
-        .gesture(TapGesture(count: 2).onEnded {
-            if let url = URL(string: app.link) {
-                downloadApp(url, app)
+        Group {
+            if #available(macOS 13.0, *) {
```
Comment (IsaacMarovitz) [inline PlayCover/Views/App Views/StoreAppView.swift]: Does the <13.0 solution work without caveats on 13.0? If so we should just stick with the one and avoid the `#available` all together

File: PlayCover/Views/Sidebar Views/IPALibraryView.swift [hunk 8]
```diff
@@ -65,11 +36,106 @@ struct IPALibraryView: View {
                     Spacer()
                 }
                 .frame(maxWidth: .infinity, maxHeight: .infinity)
+                .sheet(isPresented: $addSourcePresented) {
+                    AddSourceView(addSourceSheet: $addSourcePresented)
+                        .environmentObject(storeVM)
+                }
+            } else {
+                if #available(macOS 13.0, *) {
```
Comment (IsaacMarovitz) [inline PlayCover/Views/Sidebar Views/IPALibraryView.swift]: Again question here about the necessity of the `#available`

File: README.md [hunk 9]
```diff
@@ -127,8 +127,11 @@ Lucas Lee - playcover@lucas.icu
 
 These open source libraries were used to create this project.
 
-* [inject](https://github.com/paradiseduo/inject)
+* [Inject](https://github.com/paradiseduo/inject)
```
Comment (IsaacMarovitz) [inline README.md]: `inject` is styled as `inject` not `Inject`

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

File: PlayCover/AppInstaller/Downloader.swift (hunk 1 above)
Comment (amirsaam) [reply PlayCover/AppInstaller/Downloader.swift:72]: I'm not familiar with FileManager API, if you have can suggest how to rework it, I'll gladly accept.
So far it didn't throw issue and plus lets with this kind of approach we are letting macOS itself take care of the temporary directory and avoid unnecessary after cleanups while the proocedInstall part does actually remove the file

File: PlayCover.xcodeproj/project.pbxproj (hunk 2 above)
Comment (amirsaam) [reply PlayCover.xcodeproj/project.pbxproj:983]: If someone else does change it and be merged this change would be automatically be the same as develop branch so not a big deal

File: PlayCover/AppInstaller/Downloader.swift (hunk 3 above)
Comment (amirsaam) [reply PlayCover/AppInstaller/Downloader.swift]: I'll change the namings 👍

File: PlayCover/Utils/ViewExtenstions.swift (hunk 4 above)
Comment (amirsaam) [reply PlayCover/Utils/ViewExtenstions.swift]: In first place I didn't separated files and did it to avoid SwiftLint errors, I'll change 👍

File: PlayCover/Views/App Views/StoreAppView.swift (hunk 7 above)
Comment (amirsaam) [reply PlayCover/Views/App Views/StoreAppView.swift]: I may rework DetailedView to make the library available in macOS 12 too so if I couldn't, I'll get back to if #available parts. But for explaining, macOS 12 has onTapGesture and simultaneousGesture that you can't put in any if statement so it is necessary to wrap it in `if #available`

File: PlayCover/Views/Sidebar Views/IPALibraryView.swift (hunk 8 above)
Comment (amirsaam) [reply PlayCover/Views/Sidebar Views/IPALibraryView.swift]: same as previously answered but with the difference of navigation stack and navigation link being used in macOS 13 part that are not available in macOS 12 so we need to provide what behaviour macOS 12 should have

File: README.md (hunk 9 above)
Comment (amirsaam) [reply README.md]: I'll revert

**Comment by amirsaam** [comment]
Updated description for latest changes.

**Comment by amirsaam** [comment]
Separated Downloader (#622), Cache System (#624) and ToastView changes (#621)
